### ISSUE TITLE
SABnzbd installation fails (Debian 9 Stretch)  - no valid OpenPGP data found.

### ISSUE THREAD

**Issue by Ekynox** [liaralabs/swizzin#25]
Hi liara,
**Issue:**
The installation of SABnzbd (e.g. "box install sabnzbd") fails silently and is further reported as "installed"
This issue is new to Debian 9 (Stretch) distro.

**Troubleshooting:**
The failure is in this section of sabnzbd.sh:
```
add-apt-repository -y ppa:jcfp/sab-addons >/dev/null 2>&1
apt update >/dev/null 2>&1
```
when reproducing it manually I get the following results shown:
```
root@box1-1:~# add-apt-repository -y ppa:jcfp/sab-addons
gpg: keybox '/tmp/tmp7zrhmxqz/pubring.gpg' created
gpg: /tmp/tmp7zrhmxqz/trustdb.gpg: trustdb created
gpg: key F13930B14BB9F05F: public key "Launchpad PPA for jcfp" imported
gpg: Total number processed: 1
gpg:               imported: 1
gpg: no valid OpenPGP data found.
root@box1-1:~# apt update
[...]
Err:8 http://ppa.launchpad.net/jcfp/nobetas/ubuntu artful InRelease
  The following signatures couldn't be verified because the public key is not available: NO_PUBKEY F13930B14BB9F05F
Err:12 https://apt.sonarr.tv master InRelease
  The following signatures couldn't be verified because the public key is not available: NO_PUBKEY EBFF6B99D9B78493
Err:9 http://ppa.launchpad.net/jcfp/ppa/ubuntu artful InRelease
  The following signatures couldn't be verified because the public key is not available: NO_PUBKEY F13930B14BB9F05F
Err:11 http://ppa.launchpad.net/jcfp/sab-addons/ubuntu artful InRelease
  The following signatures couldn't be verified because the public key is not available: NO_PUBKEY F13930B14BB9F05F
Reading package lists... Done
W: GPG error: http://ppa.launchpad.net/jcfp/nobetas/ubuntu artful InRelease: The following signatures couldn't be verified because the public key is not available: NO_PUBKEY F13930B14BB9F05F
E: The repository 'http://ppa.launchpad.net/jcfp/nobetas/ubuntu artful InRelease' is not signed.
N: Updating from such a repository can't be done securely, and is therefore disabled by default.
N: See apt-secure(8) manpage for repository creation and user configuration details.
W: GPG error: https://apt.sonarr.tv master InRelease: The following signatures couldn't be verified because the public key is not available: NO_PUBKEY EBFF6B99D9B78493
E: The repository 'https://apt.sonarr.tv master InRelease' is not signed.
N: Updating from such a repository can't be done securely, and is therefore disabled by default.
N: See apt-secure(8) manpage for repository creation and user configuration details.
W: GPG error: http://ppa.launchpad.net/jcfp/ppa/ubuntu artful InRelease: The following signatures couldn't be verified because the public key is not available: NO_PUBKEY F13930B14BB9F05F
E: The repository 'http://ppa.launchpad.net/jcfp/ppa/ubuntu artful InRelease' is not signed.
N: Updating from such a repository can't be done securely, and is therefore disabled by default.
N: See apt-secure(8) manpage for repository creation and user configuration details.
W: GPG error: http://ppa.launchpad.net/jcfp/sab-addons/ubuntu artful InRelease: The following signatures couldn't be verified because the public key is not available: NO_PUBKEY F13930B14BB9F05F
E: The repository 'http://ppa.launchpad.net/jcfp/sab-addons/ubuntu artful InRelease' is not signed.
N: Updating from such a repository can't be done securely, and is therefore disabled by default.
N: See apt-secure(8) manpage for repository creation and user configuration details.


```
**Proposed solution:**
 don't know yet - not sure

**Proposed workaround:**
```
gpg --keyserver http://keyserver.ubuntu.com --recv  F13930B14BB9F05F
gpg -a --export  F13930B14BB9F05F |apt-key add -

gpg --keyserver http://keyserver.ubuntu.com --recv  EBFF6B99D9B78493
gpg -a --export  EBFF6B99D9B78493 |apt-key add -
```
This workaround worked for me. 
After this fix and the other fix proposed here https://github.com/liaralabs/swizzin/issues/24, the installation of sabnzbd via `box install sabnzbd` completed without errors.

**Comment by liaralabs** [comment]
Thanks for the thorough and detailed bug reports with workarounds and solutions.

It seems Debian can't add keys whilst adding a PPA, so I've used your workaround with a bit of modification to grab keys from the keyserver and import them manually.

`man apt-key` actually advises against using apt-key add in Debian and to import them manually into the keyring instead. Since Ubuntu and Debian should both store apt keys in `/etc/apt/trusted.gpg.d`, I have gone ahead and written a function to import them manually.

There are still a few key files not served by a public keyserver (syncthing, btsync, emby, plex and the php7.0 repository for Jessie) that are utilizing `apt-key add`. I considered importing them manually, but in order to export we need the keyid of the imported key that was just processed. While I can grab these manually and hard code them into the script, in terms of longevity of the project, I'm concerned that these parties may change their signing key in the future which could prevent the installations of these packages. As such, I've decided to continue utilizing `apt-key add` in these circumstances, while deferring to the more proper `gpg --export` method when a public keyserver is involved.
